Modules/Tag/Traits/TaggableTrait.php

```diff
@@ -66,43 +49,31 @@ public function scopeWithTag(Builder $query, $tags, $type = 'slug')
         });
     }
 
-    /**
-     * {@inheritdoc}
-     */
-    public function tags()
+    public function tags(): MorphToMany
     {
         return $this->morphToMany(static::$tagsModel, 'taggable', 'tag__tagged', 'taggable_id', 'tag_id');
     }
 
-    /**
-     * {@inheritdoc}
-     */
-    public static function createTagsModel()
+    public static function createTagsModel(): Model
     {
         return new static::$tagsModel;
     }
 
-    /**
-     * {@inheritdoc}
-     */
-    public static function allTags()
+    public static function allTags(): Builder
     {
         $instance = new static;
 
-        return $instance->createTagsModel()->with('translations')->whereNamespace($instance->getEntityClassName());
+        return self::createTagsModel()->with('translations')->where('namespace', $instance->getEntityClassName());
     }
 
-    /**
-     * {@inheritdoc}
-     */
-    public function setTags($tags, $type = 'slug')
+    public function setTags($tags, string $type = 'slug'): bool
     {
         if (empty($tags)) {
             $tags = [];
         }
 
         // Get the current entity tags
-        $entityTags = $this->tags->pluck($type)->all();
+        $entityTags = $this->tags()->get()->pluck($type)->all();
```

Comment:
Prevents a race condition when you've already added one tag, and then try to add the same tag again. This queries the database and pulls all the tags it knows about, rather than using stale data that was cached whenever the tags relationship was first loaded.